nerfstudio/model_components/losses.py

```diff
@@ -308,3 +308,234 @@ def depth_loss(
+    reduction = 0 if divisor == 0 else torch.sum(image_loss) / divisor
+
+    return reduction
+
+
+def reduction_image_based(image_loss, M):
+    """
+    mean of average of valid pixels of an image
+    Args:
+        image_loss: loss of each image
+        M: number of valid pixels of each image
+    Returns:
+        mean of average of valid pixels of an image
+    """
+    # avoid division by 0 (if M = sum(mask) = 0: image_loss = 0)
+    valid = M.nonzero()
+
+    image_loss[valid] = image_loss[valid] / M[valid]
+
+    return torch.mean(image_loss)
+
+
+def mse_loss(prediction: torch.Tensor, target: torch.Tensor, mask: torch.Tensor, reduction=reduction_batch_based):
+    """
+    mse loss from omnidata repo
+    Args:
+        prediction: predicted depth map
+        target: ground truth depth map
+        mask: mask of valid pixels
+        reduction: reduction function, either reduction_batch_based or reduction_image_based
+    Returns:
+        mse loss based on reduction function
+    """
+    M = torch.sum(mask, (1, 2))
+    res = prediction - target
+    image_loss = torch.sum(mask * res * res, (1, 2))
+
+    return reduction(image_loss, 2 * M)
+
+
+def gradient_loss(prediction: torch.Tensor, target: torch.Tensor, mask: torch.Tensor, reduction=reduction_batch_based):
+    """
+    gradient loss from omnidata repo
+    Args:
+        prediction: predicted depth map
+        target: ground truth depth map
+        mask: mask of valid pixels
+        reduction: reduction function, either reduction_batch_based or reduction_image_based
+    Returns:
+        gradient loss based on reduction function
+    """
+    M = torch.sum(mask, (1, 2))
+
+    diff = prediction - target
+    diff = torch.mul(mask, diff)
+
+    grad_x = torch.abs(diff[:, :, 1:] - diff[:, :, :-1])
+    mask_x = torch.mul(mask[:, :, 1:], mask[:, :, :-1])
+    grad_x = torch.mul(mask_x, grad_x)
+
+    grad_y = torch.abs(diff[:, 1:, :] - diff[:, :-1, :])
+    mask_y = torch.mul(mask[:, 1:, :], mask[:, :-1, :])
+    grad_y = torch.mul(mask_y, grad_y)
+
+    image_loss = torch.sum(grad_x, (1, 2)) + torch.sum(grad_y, (1, 2))
+
+    return reduction(image_loss, M)
+
+
+class MiDaSMSELoss(nn.Module):
+    """
+    data term from MiDaS paper
+    """
+
+    def __init__(self, reduction="batch-based"):
+        super().__init__()
+
+        if reduction == "batch-based":
+            self.__reduction = reduction_batch_based
+        else:
+            self.__reduction = reduction_image_based
+
+    def forward(self, prediction: torch.Tensor, target: torch.Tensor, mask: torch.Tensor):
+        """
+        Args:
+            prediction: predicted depth map
+            target: ground truth depth map
+            mask: mask of valid pixels
+        Returns:
+            mse loss based on reduction function
+        """
+        return mse_loss(prediction, target, mask, reduction=self.__reduction)
+
+
+class GradientLoss(nn.Module):
+    """
+    regularizer term from MiDaS paper
```

Comment:
Can you provide more information about what this loss does and also documentation for `scales`. Any reason not to add the scales logic to the function defined at L411 and then remove this class? It's unclear to me if there is any reason to have both a class based loss and a functional one.